vernac/metasyntax.ml

```diff
@@ -1231,6 +1238,9 @@ let find_precedence custom lev etyps symbols onlyprint =
             | _ ->
               user_err Pp.(str "A notation starting with an atomic expression must be at level 0.")
             end
+        | (ETPattern _ | ETBinder _), _ when not onlyprint ->
+            (* Don't know exactly if we can make sense of this case *)
+            user_err Pp.(str "Binders or patterns not supported in leftmost position.")
```

Comment:
Hi @herbelin, thanks for creating a patch so quickly. I have one question. I tested your branch on my development, and I'm actually using patterns quite a bit in a leftmost position. It seemed to be working quite well (except for the bugs I reported of course). What is the reason why this cannot be supported?

Reply:
I can see that this is a problem in general, but can an exception be made for custom entries?

Reply:
Now that I'm looking at this a bit more closely, it seems that the meaning of `pattern` in custom entries is actually different from normal notations. As far as I can tell, normal notations it is a [pattern](https://coq.inria.fr/refman/language/core/variants.html#grammar-token-pattern), while in custom entries it is a [pattern0](https://coq.inria.fr/refman/language/core/variants.html#grammar-token-pattern0). See the tests below. It seems to me that a `pattern0` should be quite easy to parse even in a leftmost position right?
```
Notation "[ x ]" := x (at level 0, x pattern).
Definition x {A} (y : option A) := match y with
| None => None
| [ Some x as y ] => y
end. (* Works *)
```
```
Declare Custom Entry t.
Notation "[ x ]" := x (at level 0, x custom t at level 0).
Notation "x" := x (in custom t at level 0, x pattern).

Definition x {A} (y : option A) := match y with
| None => None
| [ (Some x as y) ] => y
end. (* Works *)

Definition x {A} (y : option A) := match y with
| None => None
| [ Some x as y ] => y
end. (* Does not work *)
```

Reply:
> What is the reason why this cannot be supported?

The current status is that a notation creates two rules: one in `constr` and one in `pattern` (e.g. the notation `a :: l` for lists is both a notation in `constr` and a notation in `pattern`). 

So, if there is a notation with `pattern` in leftmost position, it will add a rule in the `pattern` grammar which refers to `pattern` itself. Of course, it would be ok if that leftmost occurrence were at a next level (so that no infinite loop happens) but since I had no use case for it, I simply forbid it.

More generally, the non-terminal `pattern` in `constr` is supposed to be used as a binder. Patterns (in `match`) are not supposed to include binders. So, I don't know what meaning to give to a notation in `pattern` which refers to a `pattern` in it. For instance, what kind of interesting application could we give to a notation for `∃ ' pattern , c` in the syntax of `pattern`?

I'd be interested by your use cases.

Reply:
My use-cases for `pattern` are two-fold:

1. I'm using `pattern` as a workaround for #9755. I need a grammar entry that can parse globals, strings and integers at the same time, while not parsing full terms. `pattern` does this quite well (although I know this is a hack).
2. My more legitimate use-case is that I need to pattern-match on long and complex expressions involving `sum, unit, option` and `pair`. This becomes much more readable when I create a small custom syntax for this (which also corresponds closely to the language I'm trying to reason about). Here is a snippet of the notations I'm using:
```
Notation "x ⃗ y .. z" := (Some (pair .. (Some (pair x y)) .. z))
  ( in custom sequence at level 0, x pattern
  , y custom singleton at level 0, z custom singleton at level 0) : fragment_scope.
Notation "x ⃗" := x
  ( in custom sequence at level 0, x pattern) : fragment_scope.
Notation "x y .. z" := (Some (pair .. (Some (pair (Some (pair None x)) y)) .. z))
  ( in custom sequence at level 0, x  pattern
  , y custom singleton at level 0, z custom singleton at level 0) : fragment_scope.
Notation "x" := (Some (pair None x))
  ( in custom sequence at level 0, x  pattern) : fragment_scope.
Notation "[ x ] y .. z" := (Some (pair .. (Some (pair (Some (pair None (inr x))) y)) .. z))
  ( in custom sequence at level 0, x at level 0
  , y custom singleton at level 0, z custom singleton at level 0) : fragment_scope.
Notation "[ x ]" := (Some (pair None (inr x)))
  ( in custom sequence at level 0, x at level 0) : fragment_scope.
Notation "[ ] y .. z" := (Some (pair .. (Some (pair (Some (pair None (inr None))) y)) .. z))
  ( in custom sequence at level 0
  , y custom singleton at level 0, z custom singleton at level 0) : fragment_scope.
Notation "[ ]" := (Some (pair None (inr None)))
  ( in custom sequence at level 0) : fragment_scope.

Notation "[ x .. y ]" := (pair .. (pair tt x) .. y)
  (x custom singleton at level 0, y custom singleton at level 0) : fragment_scope.
Notation "x" := x (in custom singleton at level 0, x pattern, only parsing) : fragment_scope.
Notation "[ k ]" := (inr k)
  (in custom singleton at level 0, k custom sequence at level 0) : fragment_scope.
Notation "[ ]" := (inr None) (in custom singleton) : fragment_scope.
Notation "[ ]" := tt : fragment_scope.
```
All of this happens at level 0, and does not enter an infinite loop nor does it throw a stackoverflow. On the contrary, once I worked around known bugs it worked quite well :-)

Reply:
>     1. I'm using `pattern` as a workaround for #9755. I need a grammar entry that can parse globals, strings and integers at the same time, while not parsing full terms. `pattern` does this quite well (although I know this is a hack).

Can't you use `constr at level 0` for this purpose?

The idea with `pattern` was that it played a role of binders. For instance, `pattern` includes an `as` clause (in the constr grammar), which makes sense mostly in position of binder.

Of course, we could make an exception and consider `pattern` as a subset of `constr` for the case of printing when it is clear that it does not play a binding role but is it clear that it is better in this case than relying on `constr`?

Reply:
So I just tried replacing (almost) all of the instances of `x pattern` with `x constr at level 0`, and this actually works fine! Now the only `pattern` I have left is for monadic do-notation:
```
Notation "'do' X <- A ; B" := (bind A (fun X => B))
  (at level 200, X pattern, A at level 100, B at level 200).
```
This is one were `pattern` is actually meant to be used right?

You'll have to excuse my ignorance. I'm doing all this just with trial and error, and I wrongly interpreted `pattern` as being for the purpose of writing pattern match expressions. The only reason why `pattern` worked for my is that it has level 0 by default. So apparently `constr` works for all of my purposes as well as long as I use level 0.

I can therefore retract all of my objections to this PR. Thanks so much for your help!

Reply:
> So I just tried replacing (almost) all of the instances of `x pattern` with `x constr at level 0`, and this actually works fine!

Great!

> Now the only `pattern` I have left is for monadic do-notation:
> 
> ```
> Notation "'do' X <- A ; B" := (bind A (fun X => B))
>   (at level 200, X pattern, A at level 100, B at level 200).
> ```
> 
> This is one were `pattern` is actually meant to be used right?

Yes, exactly.
 
Thanks for your feedback!

Reply:
Although this snippet seems to suggest that `pattern` is also apropriate in matches sometimes. Or am I stil misunderstanding things.
https://github.com/coq/coq/blob/9c2228ff011dc6188b70084fa1e1a5158affcf24/theories/Init/Notations.v#L110-L112

Reply:
Yes, `pattern` is appropriate in binding position of a `constr` and this is what happens in the `if-is` notation above.

However, a notation also defines a rule for printing `pattern` (as can be shown with `Print Grammar pattern`). Having binding patterns inside rules in the grammar of pattern is what would be difficult to explain.

The `constr` grammar is dependent on the `pattern` grammar, and it uses it to print patterns of a `match`. (But that's not reciprocal: the `pattern` grammar does not depend on `constr`.) Patterns of a `match` are purely applicative (made of constructors and variables), so we don't expect to have binders (such as a `forall`, or, recursively, a `match with end` in them).

Told in another way, `pattern` is appropriate in the pattern position of a `match`, but a `match` term (as above) is not appropriate in pattern position of a constr. Hoping that I'm not too confuse.

Reply:
Okay, most things are now starting to make sense to me (why and when we need `pattern`, and why we should't print it in `match`-statements).

However, it seems to me that you can still lift the restriction on `pattern`s in a left-most position within custom grammars. I am experimentally unable to create any kind of stackoverflow of loop when the notations are within a custom grammar. (I do not need this anymore after your suggestions though, so feel free to ignore this comment.)

Reply:
> However, it seems to me that you can still lift the restriction on `patterns` in a left-most position within custom grammars.

Yes, we can do that. I updated the PR in this direction.

Actually, the story with custom pattern notations is a bit incomplete. When a notation is declared in a custom entry `foo`, the rule is added both the custom constr entry (internally named `constr:foo`) and to a custom pattern entry (internally named `pattern:foo`). This is useful e.g. for printing `match` patterns, as in:
```
Declare Custom Entry foo.
Notation "[ x ]" := x (x custom foo at level 0).
Notation "#" := tt (in custom foo).
Check fun x => match x with tt => 0 end.
(* fun x : unit => match x with | [#] => 0 end *)
```
However, there is no syntax to give access to this custom pattern entry from another notation, nor even a way to print this entry. In particular, if one says `pattern` in a custom entry, it is a reference to the main `pattern` entry (the one associated to `constr`), not to the `pattern` entry associated to the custom grammar.

Maybe syntaxes like `x custom pattern foo` and `Print Custom Pattern Grammar foo` could respectively be used for that purposes??

Reply:
So I assume that the entry `pattern` refers to `pattern:foo` right? This actually sounds a bit problematic to me, because a notation in `pattern:foo` could have a variable `x constr` in it, which means that suddenly complete terms are allowed in `pattern`s. Or are references to `constr` somehow transformed to `pattern`?

I cannot really think of a reason why one might want to write `x custom pattern foo` instead of `x custom foo`, except maybe for printing purposes. Do you have something in mind here? I should note that for my particular use-case printing is currently pretty broken anyway due to #13042. This issue forces me to have overlapping notations that belong in different entries to share the same entry with different scopes. But while printing, the system then starts using rules from both scopes at the same time.

Reply:
> So I assume that the entry `pattern` refers to `pattern:foo` right?

The `x pattern` modifier currently always refer to the main `pattern` entry. There is currently no way to refer to the `pattern:foo`.

> This actually sounds a bit problematic to me, because a notation in `pattern:foo` could have a variable `x constr` in it, which means that suddenly complete terms are allowed in `pattern`s. Or are references to `constr` somehow transformed to `pattern`?

The latter.

> I cannot really think of a reason why one might want to write `x custom pattern foo` instead of `x custom foo`, except maybe for printing purposes. Do you have something in mind here?

Assume you want a `let '` in the custom grammar `foo`, then you would write `Notation "'let' ' p := t 'in' u" := (let 'p := t in u) (in custom foo at level 200, p pattern).", or something like that, and you would expect the pattern to be the one attached to `foo`, not the main one which goes with `constr`.

Reply:
I see, that might indeed come up when doing some kind of (P)HOAS encoding of a language. The syntax you are proposing looks reasonable to me, although it seems a bit messy that this has to exist as a corner case.